src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp

```diff
@@ -73,11 +74,11 @@ OopClosure* ShenandoahStackWatermark::closure_from_context(void* context) {
     assert(Thread::current()->is_Worker_thread(), "Unexpected thread passing in context: " PTR_FORMAT, p2i(context));
     return reinterpret_cast<OopClosure*>(context);
   } else {
-    if (_heap->is_concurrent_mark_in_progress()) {
-      return &_keep_alive_cl;
-    } else if (_heap->is_concurrent_weak_root_in_progress()) {
+    if (_heap->is_concurrent_weak_root_in_progress()) {
```

Comment:
Comprehension check: We flips this because in generational mode we _can_ have both concurrent weak roots and concurrent mark in progress at the same time, and we need to prioritize evacs, right?

Reply:
Correct. Although no old marking threads will ever run during a young collection.